Several tentative updates regarding private computation for ML, the usage of aggregator/coordinator terms and minor typos (#37)
patcg/docs-and-reports: threat-model/readme.md

In this document, we outline the security considerations for proposed purpose-constrained APIs for the web platform (that is, within browsers, mobileOSs, and other user-agents) specified by the Private Advertising Technologies Working Group (PATWG).

> Many of these proposals attempt to leverage the concept of _private computation_ as a component of these purpose-constrained APIs. An ideal private computation system would allow for the evaluation of a pre-defined function (i.e., the constrained purpose,) without revealing any new information to any party beyond the output of that predefined function. For instance, private computation could be leveraged to address two main advertising use-cases namely reporting and campaign optimization. Regarding reporting, private computation could be used to perform aggregation over inputs which, individually, must not be revealed. For campaign optimization, private computation could be used to train a machine learning algorithm over those private inputs without revealing them.

Reviewer: Do we need to condition that last statement a little more?

I say that because most of the ML proposals involve the private computation only forming a part of the training.  That is, you don't execute most of the protocol in private computation, but only those components that operate on the private data.

```suggestion
Many of these proposals attempt to leverage the concept of _private computation_ as a component of these purpose-constrained APIs. An ideal private computation system would allow for the evaluation of a pre-defined function (i.e., the constrained purpose,) without revealing any new information to any party beyond the output of that predefined function. For instance, private computation could be leveraged to address two main advertising use-cases namely reporting and campaign optimization. Regarding reporting, private computation could be used to perform aggregation over inputs which, individually, must not be revealed. For campaign optimization, private computation could be part of a system that trains a machine learning algorithm, without allowing direct access to private information.
```

Private computation can be instantiated using several technologies:

* Multi-party computation (MPC) distributes information between multiple independent entities using secret sharing.
* A trusted execution environment (TEE) isolates computation and its state by using specialized hardware.
* Fully homomorphic encryption (FHE) enables computation on the ciphertext of encrypted inputs.

> Though the implementation details differ for each technology, ultimately they all rely on finding at least two entities (also called helper parties) that can be trusted not to conspire to reveal private inputs. Such entities will be referred to as _aggregators_ and _coordinators_ for MPC-based and TEE-based private computations, respectively.

Reviewer: ```suggestion
Though the implementation details differ for each technology, ultimately they all rely on finding at least two entities that can be trusted not to conspire to reveal private inputs. Such entities will be referred to as _aggregators_ and _coordinators_ for MPC-based and TEE-based private computations, respectively.
```

> Though the implementation details differ for each technology, ultimately they all rely on finding at least two entities (also called helper parties) that can be trusted not to conspire to reveal private inputs. In the sequel, such entities will be referred to as _aggregators_ and _coordinators_ for MPC-based and TEE-based private computations, respectively.

Reviewer 2: ```suggestion
Though the implementation details differ for each technology, ultimately they all rely on finding at least two entities (also called helper parties) that can be trusted not to conspire to reveal private inputs. Such entities will be referred to as _aggregators_ and _coordinators_ for MPC-based and TEE-based private computations, respectively.
```

Reviewer 2: Not sure what is meant by "in the sequel"

Author: In the sequel means _In what follows_. It could be replaced by _In the following sections_ if more explicit.

> For our threat model, we assume that an active attacker can control a network of helper parties and has the ability to corrupt any number of clients, the parties who call the proposed APIs, and some subset of aggregators or collectors, when used.

Reviewer 2: ```suggestion
For our threat model, we assume that an active attacker can control the network and has the ability to corrupt any number of clients, the parties who call the proposed APIs, and some subset of aggregators or collectors, when used.
```

Reviewer 2: I think "the network" is correct here as we're talking about an adversary that can arbitrarily read and write network traffic

Author: Agreed.

In the presence of this adversary, APIs should aim to achieve the following goals:

> 1. **Privacy**: Clients (and, more specifically, the vendors who distribute the clients) trust that (within the threat models), the API is purpose constrained. That is, all parties receive nothing beyond the intended result (e.g., a differentially private aggregation function computed over the client inputs.)

Reviewer: I think that learning is a better choice than "receive" here.  Receive implies something far more literal.

Author: I am in favour of that if "learning" does not refer to any post-processing operation.

> 2. **Correctness:** Parties receiving the intended result trust (within the threat models) that the protocol is executed correctly. Moreover, the amount that a result can be skewed by malicious input is bounded and known.

Reviewer: This clarification is unnecessary.

> 2. **Correctness:** Parties receiving the intended result have the guarantee that the protocol is executed correctly. Moreover, the amount that a result can be skewed by malicious input is bounded and known.

Reviewer 2: I'm not sure about this one, as this puts a large burden on the protocol itself (e.g. this language might imply that the protocol must be malicious secure in its correctness). I think we should accept protocols which offer weaker notions of correctness than a 100% guarantee.  Maybe we should copy over language from the privacy section here, saying that parties trust (within the threat models) that the protocol is executed correctly?

Author: Agreed. This suggestion will be made in an additional commit. Thank you!

#### 1.2.3. Mitigations

1. Modification of client assets should be limited by the API interface to only allow for intended modifications.

> 2. Use of differential privacy (see [section 3. Aggregation and Anonymization](#3-Aggregation-and-Anonymization)) should be used to prevent [This sentence is not finished]

Reviewer: ```suggestion
2. Use of differential privacy (see [Section 3. Aggregation and Anonymization](#3-Aggregation-and-Anonymization)) should be used to protect the contributions of individual users.
```

### 1.7. Helper party collusion

If enough helper parties collude (beyond the proposal-specific subset which an attacker is assumed to control), then none of the properties of the system hold. Such scenarios are outside the threat model.

However, we do assume that an attacker can always control at least one helper party. That is, there can be no perfectly trusted helper parties.

> ### 1.8 On Premise Solutions for Helper Parties

Reviewer: It might be better to open an issue for this rather than provide only the section heading.

### 1.9 Operators of TEEs

> As a piece of hardware, TEEs will have an operator with access to the machine. Depending on the specific hardware, there may be known vulnerabilities in which an attacker who only controls the operator can violate the obliviousness of client/user data. These attacks are outside this threat model, but are likely to inform specific web platform decisions about which instantiations of private computation to support.

Reviewer 3: I appreciate this change moving away from assuming the TEE operator is a cloud provider, but without at least mentioning them as an example, I think it's hard for a reader to figure out what kind of party is being described.

> As a piece of hardware, TEEs will have an operator with access to the machine. Depending on the specific hardware, there may be known vulnerabilities in which an attacker who only controls the operator can violate the obliviousness of client/user data. These attacks are outside this threat model, but are likely to inform specific web platform decisions about which instantiations of private computation to support.

Reviewer 4: @mvono @michaelkleber Can you resolve this change before we close this PR?

> As a piece of hardware, TEEs will have an operator with access to the machine. Depending on the specific hardware, there may be known vulnerabilities in which an attacker who only controls the operator can violate the obliviousness of client/user data. These attacks are outside this threat model, but are likely to inform specific web platform decisions about which instantiations of private computation to support.

Reviewer 3: done
